examples/stable_diffusion_xl/gm/modules/lora/layers.py

```diff
@@ -0,0 +1,121 @@
+# reference to https://github.com/microsoft/LoRA
+
+import math
+import numpy as np
+
+from mindspore import nn, ops, Tensor, Parameter
+from mindspore.common import initializer as init
+from mindspore.nn.layer.basic import check_dense_input_shape
+
+
+class Identity(nn.Cell):
+    def construct(self, x):
+        return x
+
+
+class LoRALayer:
+    def __init__(
+        self,
+        r: int,
+        lora_alpha: int,
+        lora_dropout: float,
+        merge_weights: bool,
+    ):
+        assert r > 0, f"LoRA layer rank dim must be greater than 0, but got {r}."
+        self.r = r
+        self.lora_alpha = lora_alpha
+        # Optional dropout
+        if lora_dropout > 0.:
+            self.lora_dropout = nn.Dropout(p=lora_dropout)
+        else:
+            self.lora_dropout = Identity()
+
+        # Mark the weight as unmerged
+        self.merged = False
+        self.merge_weights = merge_weights
+
+
+class Dense(nn.Dense, LoRALayer):
+    # LoRA implemented in a dense layer
+    def __init__(
+        self,
+        in_features: int,
+        out_features: int,
+        r: int = 0,
+        lora_alpha: int = 1,
+        lora_dropout: float = 0.,
+        merge_weights: bool = True,
+        **kwargs
+    ):
+        nn.Dense.__init__(self, in_features, out_features, **kwargs)
+        LoRALayer.__init__(self, r=r, lora_alpha=lora_alpha, lora_dropout=lora_dropout,
+                           merge_weights=merge_weights)
+
+        # Actual trainable parameters
+        # self.lora_A = Parameter(Tensor(np.zeros((r, in_features)), self.weight.dtype))
+        # self.lora_B = Parameter(Tensor(np.zeros((out_features, r)), self.weight.dtype))
+        self.lora_A = nn.Dense(in_features, r, has_bias=False)
+        self.lora_B = nn.Dense(r, out_features, has_bias=False)
+
+        self.scaling = self.lora_alpha / self.r
```

Comment:
With default lora_alpha=1.0, ·scaling· is 0.25. But actually, we expect scaling = 1.0 by default (referring to diffusion lora).  We may set `lora_alpha: int = None,`, and set scaling to 1. if lora_alpha is None

Reply:
fixed